### DISCUSSION TITLE
API Enhancement Requests

### DISCUSSION THREAD

**Discussion by wescopeland** [RetroAchievements/RAWeb#2081] [Ideas]
**edit: please see https://github.com/RetroAchievements/RAWeb/discussions/2081#discussioncomment-15281333**

---

Is the API missing data or would you like something changed/improved to better suit your needs? Place your request here so we don't lose it.

You can also find API support in our #coders channel on Discord: https://discord.com/channels/310192285306454017/645777658319208448

**Comment by ioslife** [comment]
Getting Achievement Type on would be helpful – `API_GetUserRecentAchievements`

I pulled down RAWeb and was gonna make a change for this, assuming it would be as simple as it was to add it to getAchievementUnlocks. Alas, it looks like this change would require a change to the DB query. It looks like it should just be adding `ach.Type` to `getAchievementsEarnedBetween` in player-history.php, but not sure if changing that has any other repercussion. So I will simply keep this as a request rather than a PR 😅

**Reply by wescopeland** [reply]
Resolved by https://github.com/RetroAchievements/RAWeb/pull/2089.

**Comment by Hexadigital** [comment]
One endpoint that I would love to see, for use in things like stream overlays and progress widgets, is an endpoint that takes in a game ID and username, and returns information relating to the game and user progress.
For the game: game title, game badge image URL, game console/platform name, total points earned, total points in set
For the achievements earned: achievement title, achievement point value, achievement image URL, achievement type (if possible), and achievement unlock timestamp (preferably unix time, or including timezone, if possible)

If possible, I'd like the game ID to be optional - and if omitted, it could use the latest game ID from the user's profile (either the latest rich presence game, or the first game in the activity list on the profile)

I'm currently using that information in a stream overlay like the below image, but I think it would be less intensive on the server to have a dedicated query, since otherwise I need to use multiple endpoints to fetch and discard a lot of information that isn't needed.
![example](https://github.com/RetroAchievements/RAWeb/assets/57190472/186c41a1-d0c1-48c6-ae9b-9796f26f793b)

**Reply by LaunchBox-C-Beats** [reply]
Pretty sure you just described the getGameInfoAndUserProgress endpoint found here: https://api-docs.retroachievements.org/v1/users/get-game-info-and-user-progress.html

**Reply by Hexadigital** [reply]
Yeah, it does look like that endpoint has everything but the achievement type - at the moment I'm using that endpoint in conjunction with GetUserRecentlyPlayedGames (to figure out the latest game played) and that seems to work well.

**Reply by wescopeland** [reply]
A large portion of this is likely handled by https://github.com/RetroAchievements/RAWeb/pull/2122.

**Reply by Hexadigital** [reply]
I'd like to bump this as a separate request again - GetGameInfoAndUserProgress and GetUserRecentlyPlayedGames are both down/disabled at the moment, so a more reliable endpoint would be appreciated.

**Comment by ioslife** [comment]
Another endpoint that I mentioned (Wes said he would look into it already, just adding here to not lose track) is something that returns `filterAndJoinGames` results from https://github.com/RetroAchievements/RAWeb/blob/master/app/Platform/Services/PlayerProgressionService.php

This would allow consumers to have access to the already combined user recent progress and user site awards calls.

**Reply by wescopeland** [reply]
I believe this might pair nicely with what @Hexadigital is proposing for a new `API_GetUserGameProgression` (?) endpoint in https://github.com/RetroAchievements/RAWeb/discussions/2081#discussioncomment-7794378.

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2122

**Comment by rsubtil** [comment]
(forwarded suggestion from Discord)

One addition that I believe is really useful is an endpoint to fetch the game ID (if existing) by supplying the computed game hash (according to https://docs.retroachievements.org/Game-Identification/) instead. This would be really useful for software that's more "general" purpose, such as emulator frontends.

**Reply by wescopeland** [reply]
Will be resolved by https://github.com/RetroAchievements/RAWeb/pull/2090.

**Reply by rsubtil** [reply]
Since #2090 was closed in favor of using the `getGameList` endpoint, considering the amount of information returned, I'd like to keep my usage of this endpoint to a minimum to avoid a lot of traffic/processing on your end.

Can I reasonably assume that a game's hashes are immutable, and stay the same for a long period of time, such as a few months/years? If yes, I can keep such information cached on the client, and only make requests when there's no existing local information about a game's hash.

**Reply by wescopeland** [reply]
Yes, it is a reasonable assumption that a game's hashes tend not to change very often, often in the timespan of months or years.

However, one thing I am curious about -- how is your front-end actually hashing the games?

**Reply by rsubtil** [reply]
> However, one thing I am curious about -- how is your front-end actually hashing the games?

I use the [official hash module from rcheevos](https://github.com/RetroAchievements/rcheevos/tree/develop/src/rhash), which I had to port to Godot, available here: https://github.com/retrohub-org/godot-rcheevos-rhash

**Reply by wescopeland** [reply]
Awesome! Do you happen to be in the RetroAchievements Discord? If so, can you please DM me?

**Comment by LaunchBox-C-Beats** [comment]
Right now in LaunchBox there is a window that displays the Top Ten user leaderboard. We get the information from `GetTopTenUsers`. Right now after making that call we then have to call `GetUserSummary` 10x so that we can get those users profile pic to display in said leaderboard. Would save the number of API calls we'd have to make if that information was returned in the `GetTopTenUsers` return set.

**Comment by wescopeland** [comment]
It probably makes sense to include the user's highest award level on the game info & user progress call. Now with beaten awards introduced, this is unclear without making an additional call to get user awards.

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2122

**Comment by ioslife** [comment]
Add pagination to API_GetUserCompletedGames

Calling for a user that has a ton of records (i.e. MaxMilyin) is significantly slower (1106 ms) than a user with fewer records (i.e. Bronson15 – 263 ms)

**Reply by wescopeland** [reply]
Not sure how to do this in a way that's not a breaking change.

**Reply by ioslife** [reply]
I also didn't have an idea. Just wanted to add that it would be helpful :)

**Reply by LaunchBox-C-Beats** [reply]
This is coming from a C# dev who admittedly has little web experience, but the concept sounds doable to me. Would it be possible to add two optional parameters for page number and records per page. Page number defaults to null, records per page can default to whatever makes most sense. Then in the API logic is page number is null return the full dataset, otherwise return the number of records requested. Consumer would know to call for more pages if the number of results returned was the same as the results per page count, stop when the number of records is less than that.

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2122 will support pagination and return only 1 record per each game.

**Reply by ioslife** [reply]
Merry Christmas! Thanks for all these updates!!

**Comment by ioslife** [comment]
Add Achievement Image URL to API_GetAchievementOfTheWeek

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2121

**Comment by ioslife** [comment]
Return softcore points in API_GetAchievementUnlocks

**Reply by wescopeland** [reply]
Not sure I completely follow this request. Can you clarify? I see `Points` being returned, which is going to always be the same for both hardcore and softcore.

**Reply by ioslife** [reply]
Yeah, sorry. AchievementUnlocks returns

{
    "user": "Podgicus0305",
    "raPoints": 15544,
    "dateAwarded": "2022-07-12 19:06:34",
    "hardcoreMode": true
  }

I believe raPoints is hardcore, but I’d like to get softcore too for those that don’t play hardcore.

**Reply by ioslife** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2495

**Comment by ioslife** [comment]
Add endpoints for:
• GET and POST(!) Want to Play games
• GET and POST(!) Follow/Unfollow users

**Reply by wescopeland** [reply]
What response shapes would be desirable for these GET requests?

**Reply by ioslife** [reply]
Followed user would be nice to have something like this:
```
FollowedUser {
	username: "Bronson15",
	points: 999999,
	softcorePoints: 999999
}

GetFollowedUsers {
	[
		FollowedUser1,
		FollowedUser2,
		FollowedUser3
	]
}
```

WantToPlayGames would be nice to have something like this:
```
WantToPlayGame {
	id: 1,
	name: "Sonic the Hedgehog"",
	consoleName: "Genesis",
	imageIcon: "/url/here.png",
	maxPossible: 70,
	numAwarded: 35,
	numAwardedHardcore: 15
}

GetWantToPlayGames {
	[
		WantToPlayGame1,
		WantToPlayGame2,
		WantToPlayGame3
	]
}
```

**Reply by joamjoamjoam** [reply]
Plus one for this. It would nice to manage the want to play list through the API.

**Reply by ioslife** [reply]
> Plus one for this. It would nice to manage the want to play list through the API.

I started working on this. For now it will only be a GET request, so managing via API will not be possible. 

https://github.com/RetroAchievements/RAWeb/pull/2549

**Comment by Hexadigital** [comment]
I don't know how much interest there would be in it outside of my use case, but one thing that I was doing before (that I can't do now) was sending a POST request to reorderSiteAwards.php, in order to systematically sort my badges. If there was an API endpoint that took in an ordered list of game ID's (at a minimum) or a more complicated structure and updated the user's profile accordingly, that would be greatly appreciated.

**Comment by Nepiki** [comment]
Posted this on [Discord](https://discord.com/channels/310192285306454017/1187330228071190588) as well, but it would be ideal if the `getUserCompletedGames` endpoint could account for progression (unfinished, beaten, mastered) so we can determine a game's state on import/daily sync.

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2122

**Comment by ioslife** [comment]
Add an endpoint to return the New Sets/Revisions list

**Reply by ioslife** [reply]
This actually is accessible via the getClaims call and sorting on UpdatedDate. getClaims seems to return all claims (when do they fall off of this call?), so it would be cool to have a way to only get the most recent X amount of days

**Comment by evanation** [comment]
Hi!
Please, add 'Release' for API_GetGameList.php
and add rom name in hash table for API_GetGameList.php too

**Comment by ccreight** [comment]
Hi! I was wondering what the best way to get the image for a certain achievement was? I'm currently calling the API to get the achievement of the week, and that works alright, but I'm only able to access the achievement's ID:

achievementOfTheWeek.achievement.id

Not the actual badge's ID, to put that into a URL like this: https://i.retroachievements.org/Badge/341926.png

is there a different way I should be getting the badge image? thank you so much!

**Reply by ccreight** [reply]
ah nevermind, I found a workaround by calling for the extended game data of the related game and then finding the specific achievement's info!

**Reply by ioslife** [reply]
If you’re specifically concerned about AotW, https://github.com/RetroAchievements/RAWeb/pull/2121 is adding the imageURL to the response

**Comment by Player1041** [comment]
Would it be possible to get an endpoint with the Achievement of the Month? Basically the Achievement of the Week endpoint, but with the monthly achievement data opposed to the weekly one. 
Additionally, could you add the swaps for both weekly and monthly? :)

**Reply by wescopeland** [reply]
Unfortunately, achievement of the month is not currently persisted anywhere in the database.

**Reply by Player1041** [reply]
Ah damn. Would this be something that could be added in the future or is it somewhat out of the question?

**Comment by gilramot** [comment]
Adding a parameter of `count`, like in `API_GetUserRecentlyPlayedGames`, to `API_GetUserRecentAchievements` would be cool. You can also do it the other way around (get user recently played games by datetime). 
I request this feature as I'm trying to make a bash script to autosave the save files to a git repo and make the commit message something meaningful.
Can't do this myself, as I don't have php knowledge and I didn't dive in to the RA databases (maybe I'll learn php and pick it up if nobody'll do it).

**Comment by Nelfe80** [comment]
Hello,

I'm working on your API with Leaderboard support.

4 endpoints:
**public/API/API_GetGameInfoAndUserProgress.php** : update including leaderboards
OR/AND
**public/API/API_GetLeaderboardCount.php** : to get leaderboards from a game

**public/API/API_GetLeaderboardRanking.php** :  to get rank of the user in the specified leaderboard
**public/API/API_GetLeaderboardUsersRanking.php** : to get leaderboard scores in the specified leaderboard

Source code :
https://github.com/RetroAchievements/RAWeb/compare/master...Nelfe80:RAWeb:master

(i'm developing a marquee system to display all these retro achievements and leaderboard stats)

Thanks

Best regards

**Comment by WedgeSparda** [comment]
Hi, could it be possible to get the icon URL for each system in [All System](https://api-docs.retroachievements.org/v1/consoles/all-systems.html) response along the name and ID? Not sure if it could be possible to get it in other responses returning the Console Data, like [Achievement of the Week](https://api-docs.retroachievements.org/v1/events/achievement-of-the-week.html).

The idea behind this request is to display the console icons in a client without keeping a local image associated to the console, which can be outdated in case a new console system is added.

Thanks

**UPDATE**: I manage to make the change myself and [here](https://github.com/RetroAchievements/RAWeb/pull/2264)'s the PR

**Comment by zeroquinc** [comment]
Would love this API to include "CumulScore" (RetroPoints): https://api-docs.retroachievements.org/v1/get-user-recent-achievements.html
Would reduce an API call to the designated achievement to fetch the RetroPoints.

The by date and on date already support this.

**Reply by wescopeland** [reply]
It looks like it's being explicitly not returned in the response:

https://github.com/RetroAchievements/RAWeb/blob/master/public/API/API_GetUserRecentAchievements.php#L53

@Jamiras Would you have any reservations in us removing this `unset()` call?

**Reply by Jamiras** [reply]
`CumulScore` is literally just a summation of `Points` for the achievements being returned ([ref](https://github.com/RetroAchievements/RAWeb/blob/61d31eb1ab33c9cc95add02960d37d3c251ccb22/app/Helpers/database/player-history.php#L79-L89)).

The caller can do that summation on the client. It would be undesirable overhead to do that on the server and consume bandwidth sending that to the client.

"CumulScore" has nothing to do with RetroPoints. That's "TruePoints", and not currently returned by any of the three APIs that call `getAchievementsEarnedBetween`

**Reply by zeroquinc** [reply]
Ah I understand, well I think my question would be then: could the "TruePoints" be added to this API endpoint?

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2278

**Reply by zeroquinc** [reply]
Legend!

**Reply by zeroquinc** [reply]
Hi, I saw that it got merged. Any Idea when we can use this? Thanks!

**Reply by wescopeland** [reply]
We'll probably shoot to do a web release this Thursday.

**Comment by JeliBeanMachine** [comment]
Would very much appreciate if `API_GetGameList.php` could be modified so that the response includes the names of any supported game files instead of just the hashes.

For example, instead of:

```json
"Hashes": [
  "9a4961c50952b212666859ea4f5467a2196903ef"
]
```

you would get:

```json
"SupportedGameFiles": [
  {
    "name": "name.ext",
    "hash": "536516ce06aee351c643a3530ccb06864cae52ba"
  }
]
```

**Reply by wescopeland** [reply]
Unfortunately this would be a breaking change. This may require a new endpoint, like `API_GetGameSupportedFiles.php`.

**Reply by JeliBeanMachine** [reply]
Really, I'm just looking to retrieve a complete list of supported game files for a specified console. Making some minor edits to `API_GetGameList`, I was thinking something like this:

```php
<?php

/*
 *  API_GetSupportedFileList - returns a list of games and their supported files for the specified console
 *    i : console id
 *
 *  array
 *   object     [value]
 *    int        ID                unique identifier of the game
 *    string     Title             title of the game
 *    array      SupportedFiles
 *     object     [value]          RetroAchievements supported files associated with the game
 *      string     Hash			   The hash of the supported file
 *      string     Name			   The name of the supported file
 */

$consoleID = (int) request()->query("i");

if ($consoleID <= 0) {
    return response()->json(["success" => false]);
}

getGamesListByDev(null, $consoleID, $dataOut, 1, false, false);

$response = [];

foreach ($dataOut as &$entry) {
    $responseEntry = [
        "Title" => $entry["Title"],
        "ID" => $entry["ID"],
    ];

    $responseEntry["SupportedFiles"] = [];
    $response[] = $responseEntry;
}

foreach (getMD5List($consoleID) as $hash => $gameID) {
    foreach ($response as &$entry) {
        if ($entry["ID"] == $gameID) {
            $name = getSupportedFile($hash);
            $entry["SupportedFiles"][] = [
                "Hash" => $hash,
                "Name" => $name,
            ];
            break;
        }
    }
}

return response()->json($response);

**Reply by wescopeland** [reply]
I understand. It would definitely need to be a new endpoint -- we try to avoid breaking changes at all costs.

If you don't mind me asking, could you tell me a little more about your desired use case?

**Reply by JeliBeanMachine** [reply]
I am looking to filter my library to find files compatible with retroachievements.

**Reply by wescopeland** [reply]
Got it. One thing I am curious about -- how are you getting the hashes for your library? Are you running the rcheevos hashing algorithm under the hood?

**Reply by JeliBeanMachine** [reply]
I was using md5sum to get the hashes, but some consoles use a custom hash, in which case it might be helpful to check the file name instead.

**Reply by bbtufty** [reply]
I second this, having this would be extremely useful to expose. I have the md5 hashes from various .dat files but exposing the filenames would make the comparison simpler without having to scan the files themselves, in the cases where RA uses some custom hash. 

For files that require a patch, it would also be nice to return that as well (ideally with the download URL?)

**Comment by Gerhard-Wonner** [comment]
I would love to calculate the time-effort needed to unlock certain achievements. Therefore it would be great if I could access the time the player played a game until unlocking a certain achievement. It would generally be very interesting to access play-time related data. E. g. timestamps when a session started and when a session ended. Or the overall time a certain user played a certain game.

**Reply by wescopeland** [reply]
Unfortunately, similar to other services such as PSN and Xbox, I don't believe we'll be able to facilitate this request as there are certain privacy concerns associated with this data.

**Reply by Gerhard-Wonner** [reply]
OK, that sounds reasonable. Thank you for the quick reply!

**Comment by Gerhard-Wonner** [comment]
Please provide a function like API_GetAchievementUnlocks that returns the latest unlocks across all achievements (maybe during the last hour or with pagination). The parameters would be the same as in API_GetAchievementUnlocks, but without the target achievement ID (a). The returned data-format would be the same as the "Unlocks"-table of the response of API_GetAchievementUnlocks, but with the achievement ID as extra column.

This would allow me to keep my unlocks-database up-to-date with a lot less traffic!
Thank you! :-)

**Reply by zeroquinc** [reply]
That would be a nice addition! Would be nice to see this integrated.

**Reply by Gerhard-Wonner** [reply]
@wescopeland Have you thought about implementing this suggestion? 
It would save a lot of traffic.

**Comment by zeroquinc** [comment]
Would it be possible to have an endpoint where a user's comments are fetchable?

Usecase: I'm currently doing a Shiny only Pokemon Set and every achievement gets send to a Discord channel as embed.
I always post my SR (Soft Reset) count in the comments when I got the achievement. It would be nice to include this in the Discord message.

**Reply by ioslife** [reply]
came up with an ideal contract below https://github.com/RetroAchievements/RAWeb/discussions/2081#discussioncomment-9875779

**Comment by taras240** [comment]
It would be a good idea to add an endpoint that returns a ranked list with an offset of ±N users relative to the target, as well as the top 10 users. This way, the nearest competitors would be visible.

**Comment by ioslife** [comment]
New endpoint for retrieving comments for Games and for Achievements.

Contract would be something like:

`API_GetComments.php?z={{userName}}&y={{webAPIKey}}&t={{commentableType}}&i={{identifier}}`

`commentableType`: `[game, achievement, user]`
`identifier`: `[gameID, achievmentID, username, userID, userDisplayName]`

return:
```
{
	[
		user: 'Username',
		datePosted: 'dateTime',
		commentText: 'I am a comment, you can read me in the app now!'
	]
}
```

**Reply by ioslife** [reply]
edited after speaking with wes about contract details

**Reply by ioslife** [reply]
https://github.com/RetroAchievements/RAWeb/pull/2552

**Comment by joamjoamjoam** [comment]
I’m working on a How Long To Beat (think avg time to master based on previous masters) style display to display how long it to take to beat and master a particular achievement set. I can do this with just the achievement unlock times (no PII needed) but currently I have to get all mastery info from API_GetGameRankAndScore then query each user individually to get achievement unlock times to calculate to actual time to beat/master the set. 

Is it possible to add each users first unlock date for a set and the last unlock to beat so the total time to master/beat a set can be calculated without needing to check each user individually. 

Ideally this would be something you guys do server side  and display on the web for each set. Since you guys can look into each users unlocks for each game and filter out any long times between unlocks (user stops playing for a while) etc. maybe even have a poll for each set to vote on a difficulty rating and play through count. 

End goal:
![image](https://github.com/user-attachments/assets/7e07bd3f-af69-4ffc-8f5d-477d8cd3eb69)


Ex.
```
  {
    "User": "Arekdias",
    "NumAchievements": 15,
    "TotalScore": 219,
    "FirstAward": "2023-06-01 14:43:18"
    "LastAward": "2023-06-07 14:43:18",
    "LastAwardToBeat": "2023-06-07 14:43:18",
    "LastAwardToMaster": "2023-06-07 14:43:18"
  }
  // ...
] 
```

**Reply by joamjoamjoam** [reply]
It should be possible to map the retro ratio to a 1/10 difficulty scale too. It might be hard to map the near exponential growth rate of the RR to a useable linear scale though. Maybe some better at math can figure that one out. 

It would be nice if the API_GetGameRankAndScore returned all masteries for accurate calculations. The 10 latest masteries isn’t enough data to calculate a good average time to completion especially with some “masteries” being completed in hours or a day in games that avg much higher “time to master” times.

**Reply by thatonedood** [reply]
Its been over a year since this post, but I was just about to recommend this idea: having average time to beat/master on game pages based on previous users. Either all users can be calculated, or maybe previous 100 users. 
Just replying here to hopefully bump this idea. I would love it implemented.

**Reply by joamjoamjoam** [reply]
Im just coming back to this after a break but it looks like they added the API_GetGameProgression endpoint that shows median unlock times for achievements as well as the entire game. You can see the game averages on the beta game pages (replace game w game2 in the url).

One suggestion i have is to allow sorting games (at the console games list view level) and Achievements (at the game page view level) by average time to master/beat.

**Comment by taras240** [comment]
If possible, it would be great if the endpoint 'GetGameInfoAndUserProgress' returned the last played game when the gameID parameter is empty.

**Comment by zeroquinc** [comment]
Would it be possible that the endpoint `GetGame` returns a `PointsTotal` and `TrueRatioTotal` for the total Points and RetroPoints for that game?

**Comment by tepiloxtl** [comment]
In effect of communication on Discord, I'm putting a request for endpoint for listing set requests made by user. :)

**Reply by drisc** [reply]
I've opened a PR for this functionality here: #2950

**Comment by christianlegge** [comment]
Would be super useful for CR purposes to have a public endpoint for downloading a set's files (gameid.json, gameid-notes.json, gameid-rich.txt) without loading the game in emu.

**Reply by Jamiras** [reply]
Public users should never need direct access to any of the implementation logic. This is a niche _internal_ use case, and could reasonably use the internal APIs.

**Reply by christianlegge** [reply]
Wes said the same in discord. It'd still be nice if there was a more user-friendly way to get it than basically faking an emu-sourced request, but I understand that's way more work for you guys in basically creating a permissions framework for the API

**Comment by likeich** [comment]
I'm working on the next big update to [Console Launcher](https://play.google.com/store/apps/details?id=com.k2.consolelauncher) that has ROM frontend support. Right now I'm adding the ability to view RetroAchievements. It's been fun to add them since it makes it much more like a game console!

I have already added the achievement views for each game and console. The next feature I'm adding is viewing your followed/following list of friends, if they are active, etc. With this, it would be amazing to eventually message through the RA API as well. I would love to add a basic chat client to Console Launcher to message RA users from within the app.

|Platform with All Games | Super Mario World Cheevos |
|-------------|-------------|
| ![Screenshot_20250215-141321](https://github.com/user-attachments/assets/e36a2376-3724-450f-a8f0-d764f5091ee1) | ![Screenshot_20250216-115951](https://github.com/user-attachments/assets/643f0793-a186-4612-ae3d-fb42efd2c5d3) |

**Comment by EtherialNL** [comment]
I'm monitoring the New and Updated sets by calling the appropriate GetClaims endpoint.

According to the API documentation, about 1,000 records should be returned, but I received 11,270 records instead—with GameID 3 as the oldest—so it seems all records are being returned.

Would it be possible to add a parameter to define how many records should be returned?

The full response size (when using ClaimKind = 1) is around 4.3 MB, so limiting the number of records could help reduce bandwidth usage.

**Comment by biscuitcakes** [comment]
I'd love a way to query if an emulator is supported or not without having to launch a game. I may try to tackle it but wanted to post here anyway :)

**Comment by adamantike** [comment]
It would be great to have information regarding Hubs, like:
* Get Hub information by Hub ID, including Game IDs associated to that hub.
* List Hubs a game is part of (in the Game Extended Details endpoint, or a different one).

**Comment by turner485** [comment]
It would be great to get to have an endpoint related to the "Just Released" sets.
IE: https://retroachievements.org/claims/completed

**Reply by wescopeland** [reply]
This should be doable by using the `k` param on this endpoint: https://api-docs.retroachievements.org/v1/get-claims.html

**Reply by turner485** [reply]
Amazing thank you @wescopeland .

Is there any end point that points to this page?
https://retroachievements.org/game/33579/hashes?

**Comment by Bregann** [comment]
It'd be handy for the `API_GetUserCompletionProgress` endpoint if we could get the total points earned by the user for each game for softcore/hardcode

**Comment by m1guelpf** [comment]
I'd love to be able to access `rich_presence_updated_at` (from `PlayerSession`) in `API_GetUserSummary` or `API_GetUserData`

**Reply by m1guelpf** [reply]
Opened https://github.com/RetroAchievements/RAWeb/pull/4156 to address this

**Comment by MaXXimumBuB** [comment]
Is there something for PostComment and DeleteComment?  I would like to be able to utilize this.

Thank you.

**Reply by wescopeland** [reply]
Sorry, our API is currently read-only.

**Comment by MaXXimumBuB** [comment]
I thought so, and it makes sense, but just wanted to ask to be sure I didn’t miss it.

Thank you.

On Nov 21, 2025, at 2:46 PM, Wes Copeland ***@***.***> wrote:

﻿

Sorry, our API is currently read-only.

**Comment by Bryan1150** [comment]
I was trying to grab data from the API to make custom series/hub leaderboards for fun but I ran into 2 issues:
1. There doesn't seem to be available calls to get the data from which games are contained in a hub, this can be worked around by manually grabbing the games from the hub, but it would be nice to have a call with the hub id that returns every game in that hub
2. getGameRankAndScore only returns the top 10, so I wonder how doable it would be to have a parameter to grab it fully or atleast more than just top 10.

Sorry if those requests sound too specific for my use case, I could try my hand at implementing them if they seem to overall make sense.

**Comment by RagunovichVlad** [comment]
If possible, it would be great to add a `totalPlaytime` field to the following endpoints:

- API_GetUserRecentlyPlayedGames.php
- API_GetUserCompletionProgress.php

If that’s not feasible everywhere right now, it would be great adding it at least to:

- API_GetGameInfoAndUserProgress.php

**Reply by m1guelpf** [reply]
Opened https://github.com/RetroAchievements/RAWeb/pull/4270 for this

**Comment by wescopeland** [comment]
**Important Notice**

Once https://github.com/RetroAchievements/RAWeb/pull/4270 and https://github.com/RetroAchievements/RAWeb/pull/4210 are merged, we are placing an indefinite **code freeze** on our V1 (current) API. It will continue functioning without interruption, but all API enhancements will shift towards our new V2 web API that is under active development.

The V2 API features significantly enhanced functionality, including OAuth2 support and advanced filtering/sorting of resources. You can follow along here: https://github.com/RetroAchievements/RAWeb/pulls?q=feat%28api%29+V2

**Comment by DFelten** [comment]
Is it somehow possible to fetch the progress of an achievement? 

For example fetch the amount of trainers I've already beaten on a route of Pokémon Black for a specific achievement. The emulator is able to show me the amount. 

Or is this something which is only a part of the Savegame and not of the RA database?

**Reply by wescopeland** [reply]
Unfortunately we don't track this in the DB. It's something detected in-emulator by the achievement logic itself.

**Comment by UGamer** [comment]
Wasn't sure whether to request this now or wait until V2 is rolling, I understand y'all are hard at work on that as is!

The Achievement of the Week endpoint is awesome, and I was wondering what the likelihood of a similar endpoint for Achievement of the Month would be?

Additionally, how about one for RA Roulette? At that point maybe even a general endpoint for events where you provide the ID and it returns a list of the achievements in it.

Thank you for all that you do!

**Comment by branpg** [comment]
I want to fetch the number of set request a game has but I could only find an endpoint to get requests by user. 
I would love to have it added to the Game Extended Details if not available (So I don't have to consume more bandwith to get this).
I use my integration to the api to track my played games and I would love to use it to search for games to develop a set for. I'm currently quering my db to ask for completed games that dont have a set yet or an active claim, but I would love to be able to sort them to see wich ones have more requests.
